### ISSUE TITLE
Does not compile with gcc 13

### ISSUE THREAD

**Issue by jrmarino** [jrcarter/PragmARC#11]
apparently the new FSF compiler doesn't like this:
pragmarc-b_strings.ads:59:04: error: subtype does not statically match deferred declaration at line 27

**Comment by jrcarter** [comment]
This is a [known error](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=110488) in GNAT 13. You can stick with GNAT 12 or ignore B_Strings.

**Comment by jrmarino** [comment]
Unfortunately, dropping to gcc12 is not an option.
Can you provide a patch to "ignore B_Strings"?  otherwise I can't update the PragmArcs package which is still early 2020 version.

**Comment by jrmarino** [comment]
and is there no way to get gcc13 to compile with alternative syntax?  Yes, I understand that the code should compile as legal Ada and that you should not have to alter the code, but in the interest of getting a recent version of this packaged, is there a patch that would get B_Strings compiling?

**Comment by jrcarter** [comment]
The kind of construct that is being rejected has been legal Ada since 1983, so GNAT 13 has serious issues. The PragmARCs are portable Ada, not GNAT. I do not see any reason to impact ObjectAda users because one version of GNAT has an error.

If you wish to distribute a version of them that has been modified to compile with GNAT 13, that is certainly allowed by the license.

**Comment by jrmarino** [comment]
I understand your position and in theory I agree.
ObjectAda for me is essentially a windows product and I don't think it's a free product but I could be wrong.
so in practical terms, it should play nice with GNAT.  I agree that this regression sucks and it's a shame it's not being fixed faster.

Like I said, if you provided me a patch that would make it compile on gnat, I would simply keep the patch in the packaging repository, so it wouldn't contaminate the portable code.  I don't know offhand how to arrange the code to make GNAT happy though.

**Comment by jrcarter** [comment]
ObjectAda exists for Linux and Windows. It is a commercial compiler, though FOSS developers can get a free version, but the PragmARCs are not restricted to free compilers or free software.

A work around to the GNAT-13 error is to include the explicit discriminant

`(Max_Length => 1)`

to both declarations of the constant. I've never created a patch, so I don't know how to. Is this enough information for you to create one?

**Comment by jrmarino** [comment]
yes, that's probably enough.  I will post a link when I get this done.

**Comment by jrmarino** [comment]
so far I have not been able to work around the bug.
I send did a personal email to Eric Botcazou asking him to review the bugzilla report because I don't think anyone has seen it.

**Comment by jrmarino** [comment]
okay, this does compile.  It's the best I could do.  Apparently the bug involves the limited characteristic.
```
--- pragmarc-b_strings.ads.orig 2023-07-15 15:24:51 UTC
+++ pragmarc-b_strings.ads
@@ -21,7 +21,7 @@ pragma Assertion_Policy (Check);
 pragma Unsuppress (All_Checks);

 package PragmARC.B_Strings is
-   type B_String (Max_Length : Positive := 1024) is tagged limited private;
+   type B_String (Max_Length : Positive) is tagged private;
    -- Default initial value is Null_B_String

    Null_B_String : constant B_String; -- A string of zero characters
@@ -51,7 +51,7 @@ package PragmARC.B_Strings is
    function ">"  (Left : in B_String; Right : in B_String) return Boolean;
    function ">=" (Left : in B_String; Right : in B_String) return Boolean;
 private -- PragmARC.B_Strings
-   type B_String (Max_Length : Positive := 1024) is tagged limited record
+   type B_String (Max_Length : Positive) is tagged record
       Len   : Natural := 0;
       Value : String (1 .. Max_Length) := (1 .. Max_Length => ' ');
    end record;
```

**Comment by jrcarter** [comment]
Removing limited eliminates the essential feature of this approach: that one can declare

`B : B_String;`

and compilers that allocate the max will only allocate 1,024 bytes for the Value component.

The fix that retained limited (IIRC) was to change the constant declarations to

`Null_B_String : constant B_String (Max_length => 1); -- A string of zero characters`

and

`Null_B_String : constant B_String (Max_length => 1) := (Max_Length => 1, others => <>);`

> I send did a personal email to Eric Botcazou asking him to review the bugzilla report because I don't think anyone has seen it.

Thanks for that. I see that he did change it from "Unconfirmed" to "New" and added a target of 13.3 for a fix.

**Comment by jrmarino** [comment]
I referenced the wrong issue in my commit.
I switched the patch to your suggestion here:
https://github.com/Ravenports/ravensource/commit/47ac6807ee2811e88831d2fe9472dfc081c25fa2

**Comment by jrmarino** [comment]
FYI, it looks like the bug has been fixed in gcc 13 and the fix should be available with release 13.3.
